utils/s3/client.cc

```diff
@@ -854,6 +858,167 @@ data_sink client::make_upload_jumbo_sink(sstring object_name, std::optional<unsi
+        req.query_parameters.emplace("uploadId", _upload_id);
+        s3l.trace("PUT part {} {}~{}/{} bytes (upload id {})",
+                  part_number, total_offset, part_size, _total_size, _upload_id);
+        req.write_body("bin", part_size, [part_size, this] (output_stream<char>&& out_) mutable -> future<> {
+            auto out = std::move(out_);
+            std::exception_ptr ex;
+            try {
+                for (size_t part_offset = 0, written = 0; part_offset < part_size; part_offset += written) {
+                    auto buffer = co_await _part_body.reader.read();
+                    if (!buffer.has_value()) {
+                        // the writer side is closed, close() will abort this
+                        // multipart upload.
+                        break;
+                    }
+                    if (size_t part_remainder = part_size - part_offset; buffer->size() > part_remainder) {
+                        auto front = buffer->share(0, part_remainder);
+                        buffer->trim_front(part_remainder);
+                        // return the remainder of the buffer back to the pipe, so the next
+                        // part can pick it up
+                        _part_body.reader.unread(std::move(*buffer));
+                        co_await out.write(front.get(), front.size());
+                        written = part_remainder;
+                    } else {
+                        written = buffer->size();
+                        co_await out.write(buffer->get(), buffer->size());
+                    }
+                }
+                co_await out.flush();
+            } catch (...) {
+                ex = std::current_exception();
+            }
+            co_await out.close();
+            if (ex) {
+                std::rethrow_exception(std::move(ex));
+            }
+        });
+        auto holder = _bg_flushes.hold();
+        return _client->make_request(std::move(req), [this, part_size, part_number, start = s3_clock::now()] (group_client& gc, const http::reply& reply, input_stream<char>&& in_) mutable -> future<> {
+            auto etag = reply.get_header("ETag");
+            s3l.trace("uploaded {} part data -> etag = {} (upload id {})", part_number, etag, _upload_id);
+            _part_etags[part_number] = std::move(etag);
+            gc.write_stats.update(part_size, s3_clock::now() - start);
+            return make_ready_future();
+        }).handle_exception([this, part_number] (auto ex) {
+            s3l.warn("couldn't upload part {}: {} (upload id {})", part_number, ex, _upload_id);
+        }).finally([holder = std::move(holder)] {});
+    }
+
+    static size_t calc_part_size(size_t total_size, size_t maximum_part_size) {
+        assert(maximum_part_size >= aws_minimum_part_size);
+        for (size_t part_size = maximum_part_size; part_size >= aws_minimum_part_size; part_size -= 1_MiB) {
+            size_t num_parts = (total_size + part_size) / part_size;
```

Comment:
Shouldn't it be div-round-up? Which is `(total_size + part_size -1) / part_size`

Reply:
yeah, fixed.